---
title: Better ways in data manipulation
author: Jason West
date: '2024-03-13'
slug: better-ways-in-data-manipulation
categories:
  - Data science
  - R
tags:
  - R
  - Statistics
subtitle: 'Data wrangling with dplyr in R.'
summary: 'Data wrangling with dplyr in R.'
authors: []
lastmod: '2024-03-13T17:11:49+10:00'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---



<div id="fast-data-manipulation-with-dplyr" class="section level2">
<h2>Fast data manipulation with dplyr</h2>
<p>Hadley Wickham, author of the <a href="https://ggplot2.tidyverse.org/">ggplot2</a> library for R advances the world of computational data analysis with the <a href="http://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a> package. This package provides a grammar for data manipulation which has become as important for large-scale data analysis and manipulation as ggplot2 has become for visualization.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<p>dplyr is the next iteration of the popular plyr package. It 100 times faster and a lot more intuitive. It provides a clean interface to work with data sets that are “tidy.” Let’s use a simple analysis to show the way.</p>
<p>We analyse how much evolutionary divergence there is between human genes and the corresponding mouse orthologs. Specifically, we are interested in the range of sequence identities among genes, i.e., what are the most conserved genes, the most diverged genes, what is the mean divergence, and so on. The analysis has an additional twist in that there are different types of ortholog pairs. There are one-to-one orthologs, where the human gene has exactly one counter-part in the mouse genome, there are one-to-many orthologs, where the gene in one organism has multiple counter-parts in the other organism, and there are many-to-many orthologs, where there are multiple genes in both organisms that are orthologous to each other. So, we wish to conduct the analysis by orthology type.</p>
<p>I have obtained human-to-mouse orthologs with their respective sequence identities from <a href="http://www.ensembl.org/index.html">ensembl</a>. The resulting csv file is available <a href="https://github.com/weatherquant/data/blob/main/human_mouse_divergence.csv">here.</a> We can download this file directly into R, using the RCurl package. Loading the dplyr package is needed for fast data manipulation:</p>
<pre class="r"><code>library(RCurl)  
library(dplyr)  
url &lt;- &quot;https://raw.githubusercontent.com/weatherquant/data/main/human_mouse_divergence.csv&quot;    
data &lt;- read.csv(textConnection(getURL(url)))</code></pre>
<p>First we will take a look at the data. There are five columns, the gene id for the human gene, the gene id for the corresponding mouse ortholog, the homology type (one-to-one, one-to-many, many-to-many), the percent identity with respect to both the human (<code>perc.ident.human</code>) and the mouse (<code>perc.ident.mouse</code>) gene, and a confidence score that tells us how confident we are that the two genes are actually orthologous.</p>
<pre class="r"><code>head(data)</code></pre>
<pre><code>##     human.gene.ID      mouse.gene.ID    homology.type perc.ident.human
## 1 ENSG00000198888 ENSMUSG00000064341 ortholog_one2one          77.0440
## 2 ENSG00000198763 ENSMUSG00000064345 ortholog_one2one          57.0605
## 3 ENSG00000198804 ENSMUSG00000064351 ortholog_one2one          90.8382
## 4 ENSG00000198712 ENSMUSG00000064354 ortholog_one2one          71.3656
## 5 ENSG00000228253 ENSMUSG00000064356 ortholog_one2one          45.5882
## 6 ENSG00000198899 ENSMUSG00000064357 ortholog_one2one          75.6637
##   perc.identity.mouse confidence
## 1             77.0440          1
## 2             57.3913          1
## 3             90.6615          1
## 4             71.3656          1
## 5             46.2687          1
## 6             75.6637          1</code></pre>
<p>Now for the dplyr magic. Let’s assume we want to analyze the data by homology type, and we want to find the minimum, mean, median, and maximum sequence identity for each homology type, as well as the standard deviation of the identity distribution. We can perform this simply using the following code:</p>
<pre class="r"><code>data %&gt;% group_by(homology.type) %&gt;%  
  summarize(  
    min=min(perc.ident.human),  
    mean=mean(perc.ident.human),  
    std.dev=sd(perc.ident.human),
    max=max(perc.ident.human))</code></pre>
<pre><code>## # A tibble: 3 × 5
##   homology.type        min  mean std.dev   max
##   &lt;chr&gt;              &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
## 1 ortholog_many2many 2.13   45.8    19.8   100
## 2 ortholog_one2many  0.971  65.2    23.1   100
## 3 ortholog_one2one   2.46   85.9    12.1   100</code></pre>
<p>The function <code>group_by()</code> produces the analysis separately for each homology type, while the function <code>summarize()</code> calculates the summary statistics (min, mean, etc.) for each group. The operator <code>%&gt;%</code> is a pipe operator (in the same way as a pipe in the UNIX command line). It takes the output from the previous operation and feeds it into the subsequent operation.</p>
<p>This sequence using dplyr syntax focuses entirely on the logical flow of the data analysis. This frees us from needing to maintain a bookkeeping approach to data management, as well as avoiding the need to loop over cases, or record details related to data storage.</p>
<p>We can do plenty of related segmentation analyses quickly. Say we’re only interested in one-to-one orthologs, and we’d like to find the top 10 least diverged mouse genes and list them in descending order. The following lines achieve this very quickly:</p>
<pre class="r"><code>data %&gt;% filter(homology.type==&#39;ortholog_one2one&#39;) %&gt;%  
    select(mouse.gene.ID, human.gene.ID, perc.ident.human) %&gt;%
    arrange(desc(perc.ident.human)) %&gt;%
    do(head(.,10))</code></pre>
<pre><code>##         mouse.gene.ID   human.gene.ID perc.ident.human
## 1  ENSMUSG00000027746 ENSG00000120686              100
## 2  ENSMUSG00000027746 ENSG00000120686              100
## 3  ENSMUSG00000027746 ENSG00000120686              100
## 4  ENSMUSG00000027746 ENSG00000120686              100
## 5  ENSMUSG00000027746 ENSG00000120686              100
## 6  ENSMUSG00000021832 ENSG00000100519              100
## 7  ENSMUSG00000021832 ENSG00000100519              100
## 8  ENSMUSG00000021832 ENSG00000100519              100
## 9  ENSMUSG00000021832 ENSG00000100519              100
## 10 ENSMUSG00000021832 ENSG00000100519              100</code></pre>
<p>The function <code>filter()</code> selects all rows of the given homology type, the function <code>select()</code> picks the specific columns we are interested in, the function <code>top_n()</code> selects the top <em>n</em> values in the last data column (i.e., column <code>perc.ident.human</code> in this example), and the function <code>arrange()</code> sorts the data.</p>
<p>If these examples are of interest and you wish to learn more, the dplyr vignette offers further guidance, and can be accessed here: <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html">https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html</a></p>
<p>There is also an excellent introduction by Kevin Markham, with accompanying 40 minute video, available here: <a href="http://rpubs.com/justmarkham/dplyr-tutorial">http://rpubs.com/justmarkham/dplyr-tutorial</a></p>
<p>Finally, dplyr has a <a href="https://dbplyr.tidyverse.org/">database backend,</a> so you can now easily use all the statistical sophistication that R provides on arbitrarily large, remotely stored data sets.</p>
<p>Thanks for reading. Enjoy using this package…it’s the industry standard these days.</p>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>And if you aren’t familiar with <a href="https://ggplot2.tidyverse.org/">ggplot2,</a> spend some time with it. If you’re used to traditional visualization approaches, you may think in terms of drawing points and lines onto a canvas. ggplot2 requires you to approach visualization in a completely different way, in terms of mapping features of the data onto aesthetic features of the graph. This way of thinking is very powerful.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
